text/nodejs/0014-yarn-berry.md

```diff
@@ -0,0 +1,249 @@
+
+  [[order.group]]
+    id = "paketo-buildpacks/watchexec"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/node-engine"
+
+  [[order.group]]
+    id = "paketo-buildpacks/yarn-berry-install"
+
+  [[order.group]]
+    id = "paketo-buildpacks/node-module-bom"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/node-run-script"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/node-start"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/yarn-start"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/procfile"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/environment-variables"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/image-labels"
+    optional = true
+
+[[order]]
+
+  [[order.group]]
+    id = "paketo-buildpacks/ca-certificates"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/watchexec"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/node-engine"
+
+  [[order.group]]
+    id = "paketo-buildpacks/yarn-install"
+
+  [[order.group]]
+    id = "paketo-buildpacks/node-module-bom"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/node-run-script"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/node-start"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/yarn-start"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/procfile"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/environment-variables"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/image-labels"
+    optional = true
+
+...
+
+```
+
+The `yarn-berry-install` group is placed before the traditional `yarn-install`
+group since it detects on the presence of `.yarnrc.yml` and both app types
+will contain `yarn.lock` files.
+
+### Detection
+
+- `.yarnrc.yml` was created with the intention of making it easy to detect
+  whether a project uses Yarn Classic or Berry (Berry ignores `.yarnrc`
+  completely), and should be used for the detection of `yarn-berry-install`.
+
+### Build
+
+Yarn Berry includes a number of features which seek to remove the necessity of
+`yarn install`. As such, the buildpack should allow users to take advantage of
+these features by avoiding invocations of `yarn install` when it is reasonable
+to do so. For those who opt out of "Plug 'n Play", Berry still supports using
+`node_modules`.
+
+The buildpack should run `yarn install` in the following cases:
+
+- If an application uses `node_modules`, which may be determined by the
+  presence of a `node_modules` directory or through `.yarnrc.yml` configuration.
+
+- If there is no `.yarn/cache` in the working directory.
+
+- If the `yarn.lock` has changed.
+
+The buildpack should NOT run `yarn install` when:
+
+- Both a `.pnp.cjs` file and `.yarn/cache` are present in the working directory.
```

Comment:
Is it possible to have a `.pnp.cjs` file but no `.yarn/cache`? What about vice-versa - `.yarn/cache` but no `.pnp.cjs`? What should the buildpack do in those cases?

Reply:
In both cases, `yarn install`  would need to be run. The first case is implied above but could admittedly be made more explicit. I’ll update to include the second case as well.